File: go/basic/08-decorator/decorator.go
package decorator

import "fmt"

type wear func()

type Person interface {
	getName() string
	wear() wear
}

func gloabl_waer() {
	fmt.Println("我的着装是：")
}

type Engineer struct {
	name  string
	skill string
}

func newEngineer(name, skill string) ClothingDecorator {
	eg := Engineer{
		name:  name,
		skill: skill,
	}
	return ClothingDecorator{
		w: eg.wear(),
	}
}

func (eg *Engineer) getName() string {
	return eg.name
}

func (eg *Engineer) getSkill() string {
	return eg.skill
}

func (eg *Engineer) wear() wear {
	return func() {
		fmt.Println("我是" + eg.getSkill() + "工程师" + eg.getName())
		gloabl_waer()
	}
}

type Teacher struct {
	name  string
	title string
}

func newTeacher(name, title string) ClothingDecorator {
	tc := Teacher{
		name:  name,
		title: title,
	}
	return ClothingDecorator{
		w: tc.wear(),
	}
}

func (tc *Teacher) getName() string {
	return tc.name
}

func (tc *Teacher) getTitle() string {
	return tc.title
}

func (tc *Teacher) wear() wear {
	return func() {
		fmt.Println("我是" + tc.getTitle() + "工程师" + tc.getName())
		gloabl_waer()
	}
}

type ClothingDecorator struct {
	w wear
}

func CasualPantDecorator(ps ClothingDecorator) ClothingDecorator {
	return ClothingDecorator{
		w: func() {
			ps.w()
			fmt.Println("一条卡其色休闲裤")
		},
	}
}

func BeltDecorator(ps ClothingDecorator) ClothingDecorator {
	return ClothingDecorator{
		w: func() {
			ps.w()
			fmt.Println("一条银色针扣头的黑色腰带")
		},
	}
}

func LeatherShoesDecorator(ps ClothingDecorator) ClothingDecorator {
	return ClothingDecorator{
		w: func() {
			ps.w()
			fmt.Println("一双深色休闲皮鞋")
		},
	}
}

func KnittedSweaterDecorator(ps ClothingDecorator) ClothingDecorator {
	return ClothingDecorator{
		w: func() {
			ps.w()
			fmt.Println("一件紫红色针织毛衣")
		},
	}
}

func WhiteShirtDecorator(ps ClothingDecorator) ClothingDecorator {
	return ClothingDecorator{
		w: func() {
			ps.w()
			fmt.Println("一件白色衬衫")
		},
	}
}

func GlassesDecorator(ps ClothingDecorator) ClothingDecorator {
	return ClothingDecorator{
		w: func() {
			ps.w()
			fmt.Println("一副方形黑框眼镜")
		},
	}
}

// func main() {
// 	tony := newEngineer("Tony", "客户端")
// 	pant := CasualPantDecorator(tony)
// 	belt := BeltDecorator(pant)
// 	shoes := LeatherShoesDecorator(belt)
// 	shirt := WhiteShirtDecorator(shoes)
// 	sweater := KnittedSweaterDecorator(shirt)
// 	glasses := GlassesDecorator(sweater)
// 	glasses.w()

// 	fmt.Println()

// 	decorateTeacher := GlassesDecorator(WhiteShirtDecorator(LeatherShoesDecorator(newTeacher("wells", "教授"))))
// 	decorateTeacher.w()
// }

File: cpp/decorator/Decorator.cpp
/**
 * Author: arqqqq
 * Date: 2022/4/18
 */

#include "Decorator.h"
#include <iostream>

/* Class Person */
Person::Person(const string &name):__name(name){}

string Person::getName() const { return __name; }

void Person::wear() { printf("My wear is\n"); }

/* Class Engineer */
Engineer::Engineer(const string &name, const string &skill): Person(name), __skill(skill){}

string Engineer::getSkill() const { return __skill; }

void Engineer::wear() {
    printf("I am %s engineer %s\n", __skill.c_str(), getName().c_str());
    Person::wear();
}

/* Class Teacher */
Teacher::Teacher(const string &name, const string &title): Person(name), __title(title){}

string Teacher::getTitle() const { return __title; }

void Teacher::wear() {
    printf("I am %s %s\n", getName().c_str(), __title.c_str());
    Person::wear();
}

/* Class ClothingDecorator */
ClothingDecorator::ClothingDecorator(shared_ptr<Person> person): Person(""){
    __person = std::move(person);
}

void ClothingDecorator::wear() {
    if( __person != nullptr )
        __person->wear();
}

/* Class CasualPantDecorator */
CasualPantDecorator::CasualPantDecorator(shared_ptr<Person> person): ClothingDecorator(person){}

void CasualPantDecorator::wear() {
    ClothingDecorator::wear();
    printf("A pair of khaki slacks\n");
}

/* Class BeltDecorator */
BeltDecorator::BeltDecorator(shared_ptr<Person> person): ClothingDecorator(person){}

void BeltDecorator::wear() {
    ClothingDecorator::wear();
    printf("A black belt with silver pin buckles\n");
}

/* Class LeatherShoesDecorator */
LeatherShoesDecorator::LeatherShoesDecorator(shared_ptr<Person> person): ClothingDecorator(person){}

void LeatherShoesDecorator::wear() {
    ClothingDecorator::wear();
    printf("A pair of dark casual leather shoes\n");
}

/* Class KnittedSweaterDecorator */
KnittedSweaterDecorator::KnittedSweaterDecorator(shared_ptr<Person> person): ClothingDecorator(person){}

void KnittedSweaterDecorator::wear() {
    ClothingDecorator::wear();
    printf("A fuchsia knit sweater\n");
}

/* Class WhiteShirtDecorator */
WhiteShirtDecorator::WhiteShirtDecorator(shared_ptr<Person> person): ClothingDecorator(person){}

void WhiteShirtDecorator::wear() {
    ClothingDecorator::wear();
    printf("A white shirt\n");
}

/* Class GlassesDecorator */
GlassesDecorator::GlassesDecorator(shared_ptr<Person> person): ClothingDecorator(person){}

void GlassesDecorator::wear() {
    ClothingDecorator::wear();
    printf("A pair of square black-rimmed glasses\n");
}
